## Import Library

In [45]:
import warnings
warnings.filterwarnings('ignore')

import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

from imblearn.over_sampling import BorderlineSMOTE
from imblearn import under_sampling, over_sampling
from sklearn import preprocessing
from sklearn.impute import SimpleImputer
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, recall_score, confusion_matrix

## Load dataset

In [46]:
df = pd.read_csv('telco_customer_churn.csv')
df.sample(5)

customerID  gender  SeniorCitizen Partner Dependents  tenure  \
4325  2700-LUEVA    Male              0      No         No       1   
5162  4801-KFYKL    Male              0      No        Yes       8   
6864  5480-HPRRX  Female              1      No         No       3   
1585  4522-XRWWI    Male              0     Yes         No      42   
1400  7521-YXVZY    Male              0      No        Yes       3   

     PhoneService     MultipleLines InternetService       OnlineSecurity  ...  \
4325          Yes                No              No  No internet service  ...   
5162          Yes                No              No  No internet service  ...   
6864           No  No phone service             DSL                   No  ...   
1585          Yes               Yes             DSL                  Yes  ...   
1400          Yes                No              No  No internet service  ...   

         DeviceProtection          TechSupport          StreamingTV  \
4325  No internet service  No internet service  No internet service   
5162  No internet service  No internet service  No internet service   
6864                   No                   No                   No   
1585                   No                  Yes                  Yes   
1400  No internet service  No internet service  No internet service   

          StreamingMovies        Contract PaperlessBilling  \
4325  No internet service  Month-to-month               No   
5162  No internet service  Month-to-month               No   
6864                   No  Month-to-month               No   
1585                  Yes        One year               No   
1400  No internet service  Month-to-month               No   

                  PaymentMethod MonthlyCharges  TotalCharges Churn  
4325    Credit card (automatic)          20.75         20.75    No  
5162  Bank transfer (automatic)          19.45         159.2    No  
6864           Electronic check          25.30         77.75   Yes  
1585    Credit card (automatic)          80.45        3375.9    No  
1400               Mailed check          19.95          58.3    No  

[5 rows x 21 columns]

In [47]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


## Data Understanding

In [48]:
df.isna().sum()

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

tidak ada data null

In [6]:
df.duplicated().sum()

0

tidak ada data duplikat

In [49]:
# convert TotalCharges column to float type 
df['TotalCharges'] = df['TotalCharges'].replace(' ')
df['TotalCharges'] = df['TotalCharges'].astype('float')

In [50]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


In [51]:
categorikal = list(df.select_dtypes(include = 'object')) # filtering object column
categorikal

['customerID',
 'gender',
 'Partner',
 'Dependents',
 'PhoneService',
 'MultipleLines',
 'InternetService',
 'OnlineSecurity',
 'OnlineBackup',
 'DeviceProtection',
 'TechSupport',
 'StreamingTV',
 'StreamingMovies',
 'Contract',
 'PaperlessBilling',
 'PaymentMethod',
 'Churn']

In [52]:
for i in categorikal:
    print(f"value count for {i} is: ")
    print(df[i].value_counts(),'\n')

value count for customerID is: 
7590-VHVEG    1
3791-LGQCY    1
6008-NAIXK    1
5956-YHHRX    1
5365-LLFYV    1
             ..
9796-MVYXX    1
2637-FKFSY    1
1552-AAGRX    1
4304-TSPVK    1
3186-AJIEK    1
Name: customerID, Length: 7043, dtype: int64 

value count for gender is: 
Male      3555
Female    3488
Name: gender, dtype: int64 

value count for Partner is: 
No     3641
Yes    3402
Name: Partner, dtype: int64 

value count for Dependents is: 
No     4933
Yes    2110
Name: Dependents, dtype: int64 

value count for PhoneService is: 
Yes    6361
No      682
Name: PhoneService, dtype: int64 

value count for MultipleLines is: 
No                  3390
Yes                 2971
No phone service     682
Name: MultipleLines, dtype: int64 

value count for InternetService is: 
Fiber optic    3096
DSL            2421
No             1526
Name: InternetService, dtype: int64 

value count for OnlineSecurity is: 
No                     3498
Yes                    2019
No internet service 

## Data Preprocessing

In [53]:
# label encoding untuk value yang yes, dan no
# yes = 1, no = 0 
map_class = {'Yes': 1,
             'No': 0}
df['Partner'] = df['Partner'].map(map_class)
df['Dependents']= df['Dependents'].map(map_class)
df['PhoneService'] = df['PhoneService'].map(map_class)
df['PaperlessBilling'] = df['PaperlessBilling'].map(map_class)


In [54]:
map_gender = {'Male': 1,
             'Female': 0}
df['gender'] = df['gender'].map(map_gender)

In [55]:
df

customerID  gender  SeniorCitizen  Partner  Dependents  tenure  \
0     7590-VHVEG       0              0        1           0       1   
1     5575-GNVDE       1              0        0           0      34   
2     3668-QPYBK       1              0        0           0       2   
3     7795-CFOCW       1              0        0           0      45   
4     9237-HQITU       0              0        0           0       2   
...          ...     ...            ...      ...         ...     ...   
7038  6840-RESVB       1              0        1           1      24   
7039  2234-XADUH       0              0        1           1      72   
7040  4801-JZAZL       0              0        1           1      11   
7041  8361-LTMKD       1              1        1           0       4   
7042  3186-AJIEK       1              0        0           0      66   

      PhoneService     MultipleLines InternetService OnlineSecurity  ...  \
0                0  No phone service             DSL             No  ...   
1                1                No             DSL            Yes  ...   
2                1                No             DSL            Yes  ...   
3                0  No phone service             DSL            Yes  ...   
4                1                No     Fiber optic             No  ...   
...            ...               ...             ...            ...  ...   
7038             1               Yes             DSL            Yes  ...   
7039             1               Yes     Fiber optic             No  ...   
7040             0  No phone service             DSL            Yes  ...   
7041             1               Yes     Fiber optic             No  ...   
7042             1                No     Fiber optic            Yes  ...   

     DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0                  No          No          No              No  Month-to-month   
1                 Yes          No          No              No        One year   
2                  No          No          No              No  Month-to-month   
3                 Yes         Yes          No              No        One year   
4                  No          No          No              No  Month-to-month   
...               ...         ...         ...             ...             ...   
7038              Yes         Yes         Yes             Yes        One year   
7039              Yes          No         Yes             Yes        One year   
7040               No          No          No              No  Month-to-month   
7041               No          No          No              No  Month-to-month   
7042              Yes         Yes         Yes             Yes        Two year   

     PaperlessBilling              PaymentMethod MonthlyCharges  TotalCharges  \
0                   1           Electronic check          29.85         29.85   
1                   0               Mailed check          56.95       1889.50   
2                   1               Mailed check          53.85        108.15   
3                   0  Bank transfer (automatic)          42.30       1840.75   
4                   1           Electronic check          70.70        151.65   
...               ...                        ...            ...           ...   
7038                1               Mailed check          84.80       1990.50   
7039                1    Credit card (automatic)         103.20       7362.90   
7040                1           Electronic check          29.60        346.45   
7041                1               Mailed check          74.40        306.60   
7042                1  Bank transfer (automatic)         105.65       6844.50   

      Churn  
0        No  
1        No  
2       Yes  
3        No  
4       Yes  
...     ...  
7038     No  
7039     No  
7040     No  
7041    Yes  
7042     No  

[7043 rows x 21 columns]

Ordinal encoding kolom contract

In [56]:
map_contract = {'Month-to-month': 1,
             'One year': 2,
             'Two year': 3}
df['Contract'] = df['Contract'].map(map_contract)
df

customerID  gender  SeniorCitizen  Partner  Dependents  tenure  \
0     7590-VHVEG       0              0        1           0       1   
1     5575-GNVDE       1              0        0           0      34   
2     3668-QPYBK       1              0        0           0       2   
3     7795-CFOCW       1              0        0           0      45   
4     9237-HQITU       0              0        0           0       2   
...          ...     ...            ...      ...         ...     ...   
7038  6840-RESVB       1              0        1           1      24   
7039  2234-XADUH       0              0        1           1      72   
7040  4801-JZAZL       0              0        1           1      11   
7041  8361-LTMKD       1              1        1           0       4   
7042  3186-AJIEK       1              0        0           0      66   

      PhoneService     MultipleLines InternetService OnlineSecurity  ...  \
0                0  No phone service             DSL             No  ...   
1                1                No             DSL            Yes  ...   
2                1                No             DSL            Yes  ...   
3                0  No phone service             DSL            Yes  ...   
4                1                No     Fiber optic             No  ...   
...            ...               ...             ...            ...  ...   
7038             1               Yes             DSL            Yes  ...   
7039             1               Yes     Fiber optic             No  ...   
7040             0  No phone service             DSL            Yes  ...   
7041             1               Yes     Fiber optic             No  ...   
7042             1                No     Fiber optic            Yes  ...   

     DeviceProtection TechSupport StreamingTV StreamingMovies Contract  \
0                  No          No          No              No        1   
1                 Yes          No          No              No        2   
2                  No          No          No              No        1   
3                 Yes         Yes          No              No        2   
4                  No          No          No              No        1   
...               ...         ...         ...             ...      ...   
7038              Yes         Yes         Yes             Yes        2   
7039              Yes          No         Yes             Yes        2   
7040               No          No          No              No        1   
7041               No          No          No              No        1   
7042              Yes         Yes         Yes             Yes        3   

      PaperlessBilling              PaymentMethod MonthlyCharges  \
0                    1           Electronic check          29.85   
1                    0               Mailed check          56.95   
2                    1               Mailed check          53.85   
3                    0  Bank transfer (automatic)          42.30   
4                    1           Electronic check          70.70   
...                ...                        ...            ...   
7038                 1               Mailed check          84.80   
7039                 1    Credit card (automatic)         103.20   
7040                 1           Electronic check          29.60   
7041                 1               Mailed check          74.40   
7042                 1  Bank transfer (automatic)         105.65   

      TotalCharges  Churn  
0            29.85     No  
1          1889.50     No  
2           108.15    Yes  
3          1840.75     No  
4           151.65    Yes  
...            ...    ...  
7038       1990.50     No  
7039       7362.90     No  
7040        346.45     No  
7041        306.60    Yes  
7042       6844.50     No  

[7043 rows x 21 columns]

Frekuensi Encoding

In [57]:
freq_multilines = df['MultipleLines'].value_counts().reset_index()
freq_multilines.rename(columns = {'index':'MultipleLines','MultipleLines':'freq_multi'}, inplace = True)
freq_multilines['persen_value_multi'] = round(freq_multilines['freq_multi']/freq_multilines['freq_multi'].sum()*100,2)
freq_multilines

MultipleLines  freq_multi  persen_value_multi
0                No        3390               48.13
1               Yes        2971               42.18
2  No phone service         682                9.68

In [58]:
freq_inter_serve = df['InternetService'].value_counts().reset_index()
freq_inter_serve.rename(columns = {'index':'InternetService','InternetService':'freq_inter_serve'}, inplace = True)
freq_inter_serve['persen_inter_serve'] = round(freq_inter_serve['freq_inter_serve']/freq_inter_serve['freq_inter_serve'].sum()*100,2)
freq_inter_serve

InternetService  freq_inter_serve  persen_inter_serve
0     Fiber optic              3096               43.96
1             DSL              2421               34.37
2              No              1526               21.67

In [59]:
freq_security = df['OnlineSecurity'].value_counts().reset_index()
freq_security.rename(columns = {'index':'OnlineSecurity','OnlineSecurity':'freq_security'}, inplace = True)
freq_security['persen_security'] = round(freq_security['freq_security']/freq_security['freq_security'].sum()*100,2)
freq_security

OnlineSecurity  freq_security  persen_security
0                   No           3498            49.67
1                  Yes           2019            28.67
2  No internet service           1526            21.67

In [60]:
freq_backup = df['OnlineBackup'].value_counts().reset_index()
freq_backup.rename(columns = {'index':'OnlineBackup','OnlineBackup':'freq_backup'}, inplace = True)
freq_backup['persen_backup'] = round(freq_backup['freq_backup']/freq_backup['freq_backup'].sum()*100,2)
freq_backup

OnlineBackup  freq_backup  persen_backup
0                   No         3088          43.84
1                  Yes         2429          34.49
2  No internet service         1526          21.67

In [61]:
freq_protect = df['DeviceProtection'].value_counts().reset_index()
freq_protect.rename(columns = {'index':'DeviceProtection','DeviceProtection':'freq_protect'}, inplace = True)
freq_protect['persen_protect'] = round(freq_protect['freq_protect']/freq_protect['freq_protect'].sum()*100,2)
freq_protect

DeviceProtection  freq_protect  persen_protect
0                   No          3095           43.94
1                  Yes          2422           34.39
2  No internet service          1526           21.67

In [62]:
freq_tech_support = df['TechSupport'].value_counts().reset_index()
freq_tech_support.rename(columns = {'index':'TechSupport','TechSupport':'freq_tech_support'}, inplace = True)
freq_tech_support['persen_tech_support'] = round(freq_tech_support['freq_tech_support']/freq_tech_support['freq_tech_support'].sum()*100,2)
freq_tech_support

TechSupport  freq_tech_support  persen_tech_support
0                   No               3473                49.31
1                  Yes               2044                29.02
2  No internet service               1526                21.67

In [63]:
freq_stream_tv = df['StreamingTV'].value_counts().reset_index()
freq_stream_tv.rename(columns = {'index':'StreamingTV','StreamingTV':'freq_stream_tv'}, inplace = True)
freq_stream_tv['persen_stream_tv'] = round(freq_stream_tv['freq_stream_tv']/freq_stream_tv['freq_stream_tv'].sum()*100,2)
freq_stream_tv

StreamingTV  freq_stream_tv  persen_stream_tv
0                   No            2810             39.90
1                  Yes            2707             38.44
2  No internet service            1526             21.67

In [64]:
freq_stream_mv = df['StreamingMovies'].value_counts().reset_index()
freq_stream_mv.rename(columns = {'index':'StreamingMovies','StreamingMovies':'freq_stream_mv'}, inplace = True)
freq_stream_mv['persen_stream_mv'] = round(freq_stream_mv['freq_stream_mv']/freq_stream_mv['freq_stream_mv'].sum()*100,2)
freq_stream_mv

StreamingMovies  freq_stream_mv  persen_stream_mv
0                   No            2785             39.54
1                  Yes            2732             38.79
2  No internet service            1526             21.67

In [65]:
df = df.merge(freq_multilines[['MultipleLines','persen_value_multi']], on='MultipleLines')
df = df.merge(freq_inter_serve[['InternetService','persen_inter_serve']], on='InternetService')
df = df.merge(freq_security[['OnlineSecurity','persen_security']], on='OnlineSecurity')
df = df.merge(freq_backup[['OnlineBackup','persen_backup']], on='OnlineBackup')
df = df.merge(freq_protect[['DeviceProtection','persen_protect']], on='DeviceProtection')
df = df.merge(freq_tech_support[['TechSupport','persen_tech_support']], on='TechSupport')
df = df.merge(freq_stream_tv[['StreamingTV','persen_stream_tv']], on='StreamingTV')
df = df.merge(freq_stream_mv[['StreamingMovies','persen_stream_mv']], on='StreamingMovies')
df

customerID  gender  SeniorCitizen  Partner  Dependents  tenure  \
0     7590-VHVEG       0              0        1           0       1   
1     8665-UTDHZ       1              0        1           1       1   
2     6317-YPKDH       0              0        0           0       1   
3     6235-VDHOM       0              1        0           0       5   
4     6543-CPZMK       1              0        1           1       9   
...          ...     ...            ...      ...         ...     ...   
7038  4707-MAXGU       1              0        1           0      72   
7039  4534-WGCIR       0              0        1           1      58   
7040  6976-BWGLQ       0              0        1           1      72   
7041  2511-ALLCS       0              0        1           1      35   
7042  0804-YGEQV       0              0        1           1      43   

      PhoneService     MultipleLines InternetService       OnlineSecurity  \
0                0  No phone service             DSL                   No   
1                0  No phone service             DSL                   No   
2                0  No phone service             DSL                   No   
3                0  No phone service             DSL                   No   
4                0  No phone service             DSL                   No   
...            ...               ...             ...                  ...   
7038             1               Yes              No  No internet service   
7039             1               Yes              No  No internet service   
7040             1               Yes              No  No internet service   
7041             1               Yes              No  No internet service   
7042             1               Yes              No  No internet service   

      ... TotalCharges Churn persen_value_multi persen_inter_serve  \
0     ...        29.85    No               9.68              34.37   
1     ...        30.20   Yes               9.68              34.37   
2     ...        29.95   Yes               9.68              34.37   
3     ...       131.05   Yes               9.68              34.37   
4     ...       248.95   Yes               9.68              34.37   
...   ...          ...   ...                ...                ...   
7038  ...      1872.20    No              42.18              21.67   
7039  ...      1509.90    No              42.18              21.67   
7040  ...      1787.35    No              42.18              21.67   
7041  ...       821.60    No              42.18              21.67   
7042  ...       993.15    No              42.18              21.67   

     persen_security  persen_backup  persen_protect persen_tech_support  \
0              49.67          34.49           43.94               49.31   
1              49.67          34.49           43.94               49.31   
2              49.67          34.49           43.94               49.31   
3              49.67          34.49           43.94               49.31   
4              49.67          34.49           43.94               49.31   
...              ...            ...             ...                 ...   
7038           21.67          21.67           21.67               21.67   
7039           21.67          21.67           21.67               21.67   
7040           21.67          21.67           21.67               21.67   
7041           21.67          21.67           21.67               21.67   
7042           21.67          21.67           21.67               21.67   

      persen_stream_tv  persen_stream_mv  
0                39.90             39.54  
1                39.90             39.54  
2                39.90             39.54  
3                39.90             39.54  
4                39.90             39.54  
...                ...               ...  
7038             21.67             21.67  
7039             21.67             21.67  
7040             21.67             21.67  
7041             21.67             21.67  
7

In [66]:
df = df.drop(['MultipleLines',
 'InternetService',
 'OnlineSecurity',
 'OnlineBackup',
 'DeviceProtection',
 'TechSupport',
 'StreamingTV',
 'StreamingMovies'], axis=1)

In [67]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   customerID           7043 non-null   object 
 1   gender               7043 non-null   int64  
 2   SeniorCitizen        7043 non-null   int64  
 3   Partner              7043 non-null   int64  
 4   Dependents           7043 non-null   int64  
 5   tenure               7043 non-null   int64  
 6   PhoneService         7043 non-null   int64  
 7   Contract             7043 non-null   int64  
 8   PaperlessBilling     7043 non-null   int64  
 9   PaymentMethod        7043 non-null   object 
 10  MonthlyCharges       7043 non-null   float64
 11  TotalCharges         7043 non-null   float64
 12  Churn                7043 non-null   object 
 13  persen_value_multi   7043 non-null   float64
 14  persen_inter_serve   7043 non-null   float64
 15  persen_security      7043 non-null   f

One hot encooding payment method

In [68]:
dumies_payment_method = pd.get_dummies(df['PaymentMethod'],prefix='PaymentMethod')
dumies_payment_method

PaymentMethod_Bank transfer (automatic)  \
0                                           0   
1                                           0   
2                                           1   
3                                           0   
4                                           0   
...                                       ...   
7038                                        0   
7039                                        0   
7040                                        1   
7041                                        1   
7042                                        1   

      PaymentMethod_Credit card (automatic)  PaymentMethod_Electronic check  \
0                                         0                               1   
1                                         0                               1   
2                                         0                               0   
3                                         0                               1   
4                                         0                               1   
...                                     ...                             ...   
7038                                      1                               0   
7039                                      1                               0   
7040                                      0                               0   
7041                                      0                               0   
7042                                      0                               0   

      PaymentMethod_Mailed check  
0                              0  
1                              0  
2                              0  
3                              0  
4                              0  
...                          ...  
7038                           0  
7039                           0  
7040                           0  
7041                           0  
7042                           0  

[7043 rows x 4 columns]

In [69]:
df = pd.concat([df,dumies_payment_method], axis =1)
df

customerID  gender  SeniorCitizen  Partner  Dependents  tenure  \
0     7590-VHVEG       0              0        1           0       1   
1     8665-UTDHZ       1              0        1           1       1   
2     6317-YPKDH       0              0        0           0       1   
3     6235-VDHOM       0              1        0           0       5   
4     6543-CPZMK       1              0        1           1       9   
...          ...     ...            ...      ...         ...     ...   
7038  4707-MAXGU       1              0        1           0      72   
7039  4534-WGCIR       0              0        1           1      58   
7040  6976-BWGLQ       0              0        1           1      72   
7041  2511-ALLCS       0              0        1           1      35   
7042  0804-YGEQV       0              0        1           1      43   

      PhoneService  Contract  PaperlessBilling              PaymentMethod  \
0                0         1                 1           Electronic check   
1                0         1                 0           Electronic check   
2                0         1                 0  Bank transfer (automatic)   
3                0         1                 0           Electronic check   
4                0         1                 1           Electronic check   
...            ...       ...               ...                        ...   
7038             1         3                 0    Credit card (automatic)   
7039             1         3                 0    Credit card (automatic)   
7040             1         3                 0  Bank transfer (automatic)   
7041             1         3                 0  Bank transfer (automatic)   
7042             1         2                 1  Bank transfer (automatic)   

      ...  persen_security  persen_backup persen_protect  persen_tech_support  \
0     ...            49.67          34.49          43.94                49.31   
1     ...            49.67          34.49          43.94                49.31   
2     ...            49.67          34.49          43.94                49.31   
3     ...            49.67          34.49          43.94                49.31   
4     ...            49.67          34.49          43.94                49.31   
...   ...              ...            ...            ...                  ...   
7038  ...            21.67          21.67          21.67                21.67   
7039  ...            21.67          21.67          21.67                21.67   
7040  ...            21.67          21.67          21.67                21.67   
7041  ...            21.67          21.67          21.67                21.67   
7042  ...            21.67          21.67          21.67                21.67   

      persen_stream_tv  persen_stream_mv  \
0                39.90             39.54   
1                39.90             39.54   
2                39.90             39.54   
3                39.90             39.54   
4                39.90             39.54   
...                ...               ...   
7038             21.67             21.67   
7039             21.67             21.67   
7040             21.67             21.67   
7041             21.67             21.67   
7042             21.67             21.67   

      PaymentMethod_Bank transfer (automatic)  \
0                                           0   
1                                           0   
2                                           1   
3                                           0   
4                                           0   
...                                       ...   
7038                                        0   
7039                                        0   
7040                                        1   
7041                                        1   
7042                                        1   

      PaymentMethod_Credit card (automatic)  PaymentMethod_Electronic check  \
0                                         0   

In [70]:
df = df.drop(['PaymentMethod'], axis =1)
df

customerID  gender  SeniorCitizen  Partner  Dependents  tenure  \
0     7590-VHVEG       0              0        1           0       1   
1     8665-UTDHZ       1              0        1           1       1   
2     6317-YPKDH       0              0        0           0       1   
3     6235-VDHOM       0              1        0           0       5   
4     6543-CPZMK       1              0        1           1       9   
...          ...     ...            ...      ...         ...     ...   
7038  4707-MAXGU       1              0        1           0      72   
7039  4534-WGCIR       0              0        1           1      58   
7040  6976-BWGLQ       0              0        1           1      72   
7041  2511-ALLCS       0              0        1           1      35   
7042  0804-YGEQV       0              0        1           1      43   

      PhoneService  Contract  PaperlessBilling  MonthlyCharges  ...  \
0                0         1                 1           29.85  ...   
1                0         1                 0           30.20  ...   
2                0         1                 0           29.95  ...   
3                0         1                 0           28.45  ...   
4                0         1                 1           29.95  ...   
...            ...       ...               ...             ...  ...   
7038             1         3                 0           25.85  ...   
7039             1         3                 0           25.15  ...   
7040             1         3                 0           25.20  ...   
7041             1         3                 0           24.30  ...   
7042             1         2                 1           24.45  ...   

      persen_security persen_backup  persen_protect  persen_tech_support  \
0               49.67         34.49           43.94                49.31   
1               49.67         34.49           43.94                49.31   
2               49.67         34.49           43.94                49.31   
3               49.67         34.49           43.94                49.31   
4               49.67         34.49           43.94                49.31   
...               ...           ...             ...                  ...   
7038            21.67         21.67           21.67                21.67   
7039            21.67         21.67           21.67                21.67   
7040            21.67         21.67           21.67                21.67   
7041            21.67         21.67           21.67                21.67   
7042            21.67         21.67           21.67                21.67   

      persen_stream_tv  persen_stream_mv  \
0                39.90             39.54   
1                39.90             39.54   
2                39.90             39.54   
3                39.90             39.54   
4                39.90             39.54   
...                ...               ...   
7038             21.67             21.67   
7039             21.67             21.67   
7040             21.67             21.67   
7041             21.67             21.67   
7042             21.67             21.67   

      PaymentMethod_Bank transfer (automatic)  \
0                                           0   
1                                           0   
2                                           1   
3                                           0   
4                                           0   
...                                       ...   
7038                                        0   
7039                                        0   
7040                                        1   
7041                                        1   
7042                                        1   

      PaymentMethod_Credit card (automatic)  PaymentMethod_Electronic check  \
0                                         0                               1   
1                                         0                               1   
2                   

In [71]:
df = df.drop(['customerID'], axis =1)
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7043 entries, 0 to 7042
Data columns (total 23 columns):
 #   Column                                   Non-Null Count  Dtype  
---  ------                                   --------------  -----  
 0   gender                                   7043 non-null   int64  
 1   SeniorCitizen                            7043 non-null   int64  
 2   Partner                                  7043 non-null   int64  
 3   Dependents                               7043 non-null   int64  
 4   tenure                                   7043 non-null   int64  
 5   PhoneService                             7043 non-null   int64  
 6   Contract                                 7043 non-null   int64  
 7   PaperlessBilling                         7043 non-null   int64  
 8   MonthlyCharges                           7043 non-null   float64
 9   TotalCharges                             7043 non-null   float64
 10  Churn                                    7043 no

## Modeling (Gunakan lebih min 2 model dan bandingkan hasil evaluasinya)

Bebas menggunakan model, mau menggunakan decision tree, random forest, xgboost, dll juga boleh<br><br>
silahkan berekspresi :)

In [72]:
df['Churn'].value_counts()

No     5174
Yes    1869
Name: Churn, dtype: int64

## imbalance data handling

In [116]:
# menghapus kolom MonthlyCharges, karena punya hubungan multicolinearity dengan totalcharges
x = df.drop(['Churn','MonthlyCharges'], axis=1)
y = df['Churn']

#Over sampling with SMOTE
X_over_smote, y_over_smote = over_sampling.SMOTE().fit_resample(x,y)

In [117]:
df_oversampling_smote = pd.concat([X_over_smote, y_over_smote], axis = 1)
df_oversampling_smote

gender  SeniorCitizen  Partner  Dependents  tenure  PhoneService  \
0           0              0        1           0       1             0   
1           1              0        1           1       1             0   
2           0              0        0           0       1             0   
3           0              1        0           0       5             0   
4           1              0        1           1       9             0   
...       ...            ...      ...         ...     ...           ...   
10343       1              0        0           0       1             1   
10344       0              0        0           0       1             1   
10345       1              0        1           0      70             1   
10346       1              0        0           0       1             1   
10347       1              0        0           0      29             1   

       Contract  PaperlessBilling  TotalCharges  persen_value_multi  ...  \
0             1                 1     29.850000            9.680000  ...   
1             1                 0     30.200000            9.680000  ...   
2             1                 0     29.950000            9.680000  ...   
3             1                 0    131.050000            9.680000  ...   
4             1                 1    248.950000            9.680000  ...   
...         ...               ...           ...                 ...  ...   
10343         1                 0     79.930604           48.130000  ...   
10344         1                 1     45.616400           48.130000  ...   
10345         1                 1   7537.987673           46.653483  ...   
10346         1                 1     69.944554           48.130000  ...   
10347         1                 0   2341.827586           48.130000  ...   

       persen_backup  persen_protect  persen_tech_support  persen_stream_tv  \
0          34.490000       43.940000            49.310000         39.900000   
1          34.490000       43.940000            49.310000         39.900000   
2          34.490000       43.940000            49.310000         39.900000   
3          34.490000       43.940000            49.310000         39.900000   
4          34.490000       43.940000            49.310000         39.900000   
...              ...             ...                  ...               ...   
10343      43.840000       43.940000            49.310000         39.900000   
10344      43.840000       43.940000            49.310000         39.900000   
10345      34.490000       36.759872            34.055047         38.440000   
10346      43.840000       43.940000            49.310000         39.900000   
10347      39.333883       39.337496            49.310000         39.196371   

       persen_stream_mv  PaymentMethod_Bank transfer (automatic)  \
0                 39.54                                        0   
1                 39.54                                        0   
2                 39.54                                        1   
3                 39.54                                        0   
4                 39.54                                        0   
...                 ...                                      ...   
10343             38.79                                        0   
10344             39.54                                        0   
10345             38.79                                        0   
10346             39.54                                        0   
10347             39.54                                        0   

       PaymentMethod_Credit card (automatic)  PaymentMethod_Electronic check  \
0                                          0                               1   
1                                          0                               1   
2                                          0                               0   
3                                          0                               1   
4                          

In [128]:
df_oversampling_smote['Churn'].value_counts()

No     5174
Yes    5174
Name: Churn, dtype: int64

## feature and target

In [118]:
x = df_oversampling_smote.drop(['Churn'], axis =1)
y = df_oversampling_smote['Churn']

## scaling

In [119]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
scaler.fit(x)
X_std = scaler.transform(x)

## split data

In [120]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X_std,
                                                  y,
                                                  test_size=0.3,
                                                  random_state = 42)

# Modeling

pemodelan kali ini, saya menggunakan Logregression, random forest dan decision tree. kemudian membandingkan performa ketiganya

#### Logistic Regression

In [121]:
from sklearn.linear_model import LogisticRegression

logreg = LogisticRegression(random_state = 42).fit(X_train,y_train)

In [122]:
y_pred_logreg = logreg.predict(X_test)

In [123]:
from sklearn.metrics import classification_report 

print(classification_report(y_test,y_pred_logreg))

              precision    recall  f1-score   support

          No       0.84      0.80      0.82      1584
         Yes       0.80      0.84      0.82      1521

    accuracy                           0.82      3105
   macro avg       0.82      0.82      0.82      3105
weighted avg       0.82      0.82      0.82      3105



#### Decision Tree

In [124]:
from sklearn.tree import DecisionTreeClassifier

de_tree = DecisionTreeClassifier(random_state = 42).fit(X_train,y_train)
y_pred_decision_tree = de_tree.predict(X_test)
y_pred_decision_tree

array(['Yes', 'No', 'No', ..., 'No', 'No', 'No'], dtype=object)

In [125]:
print(classification_report(y_test,y_pred_decision_tree))

              precision    recall  f1-score   support

          No       0.81      0.79      0.80      1584
         Yes       0.78      0.80      0.79      1521

    accuracy                           0.80      3105
   macro avg       0.80      0.80      0.80      3105
weighted avg       0.80      0.80      0.80      3105



#### Random Forest

In [126]:
random_forest = RandomForestClassifier().fit(X_train, y_train)
y_pred_forest = random_forest.predict(X_test)

In [127]:
print(classification_report(y_test,y_pred_forest))

              precision    recall  f1-score   support

          No       0.84      0.85      0.85      1584
         Yes       0.84      0.83      0.84      1521

    accuracy                           0.84      3105
   macro avg       0.84      0.84      0.84      3105
weighted avg       0.84      0.84      0.84      3105



## Evaluation

pilih model yang terbaik performannya kemudian beri pejelasan kenapa model tersebut lebih baik dibandingkan dengan yang lain

model yang saya pilih adalah random_forest, karena memiliki performa paling tinggi dari model yang lain, dimana nilai akurasi f1-scorenya 0.84 untuk prediksi customer yang churn (yes), dan memiliki nilai performa precision (yes) yang paling besar